### PR TITLE
Add two-factor authentication

### LINKED ISSUE #13703: Add support for two-factor authentication
### Is your feature request related to a problem? Please describe.
Add support for two-factor authentication

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Fix #13703

2fa-settings

### CODE REVIEW

**Review by hishamco — CHANGES_REQUESTED** [review]
I did a quick review, suggesting few changes, I will revise the PR again

Great work Mike!!

**Comment by hishamco** [comment]
BTW can we make the Token length a textbox instead of forcing user to select 6 or 8 digits?

**Comment by hishamco** [comment]
@MikeAlhayek could you please provide a screenshot for asking for the token and possibly how to recover the missing token? Also it would be nice if you can demo it in the OC steering meeting

**Comment by MikeAlhayek** [comment]
It was demonstrated yesterday during the meeting :).

**Comment by MikeAlhayek** [comment]
> BTW can we make the Token length a textbox instead of forcing user to select 6 or 8 digits?

No. The specification of the Authenticator apps only support 6 or 8.

**Comment by hishamco** [comment]
I expect something like that :) unfortunately, I didn't attend last few meetings

File: src/OrchardCore/OrchardCore.Users.Core/Services/LoginSettingsExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        {
+            return settings.EnableTwoFactorAuthentication
+                || settings.EnableTwoFactorAuthenticationForSpecificRoles;
+        }
+
+        public static async Task<bool> CanEnableTwoFactorAuthenticationAsync(this LoginSettings loginSettings, Func<string, Task<bool>> isInRole)
+        {
+            if (loginSettings.EnableTwoFactorAuthenticationForSpecificRoles)
+            {
+                foreach (var role in loginSettings.Roles)
```
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Users.Core/Services/LoginSettingsExtensions.cs:18]: This is a different behavior. We only want to return true if the isInRole is true.

File: src/OrchardCore.Modules/OrchardCore.Users/Drivers/RoleLoginSettingsDisplayDriver.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                IsSelected = settings.Roles != null && settings.Roles.Contains(role.RoleName),
+            }).OrderBy(entry => entry.Role)
+            .ToArray();
+        }).Location("Content:5.1#Two-factor Authentication")
+        .OnGroup(GroupId);
+    }
+
+    public override async Task<IDisplayResult> UpdateAsync(LoginSettings section, BuildEditorContext context)
+    {
+        if (context.GroupId != GroupId || !await _authorizationService.AuthorizeAsync(_httpContextAccessor.HttpContext?.User, CommonPermissions.ManageUsers))
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/Drivers/RoleLoginSettingsDisplayDriver.cs:63]: case insensitive

File: src/OrchardCore.Modules/OrchardCore.Users/Startup.cs [hunk 3]
```diff
@@ -155,17 +208,31 @@ public override void ConfigureServices(IServiceCollection services)
...
+            });
+
+            var dataProtectionProviderType = typeof(DataProtectorTokenProvider<>).MakeGenericType(identityBuilder.UserType);
+            //var phoneNumberProviderType = typeof(PhoneNumberTokenProvider<>).MakeGenericType(identityBuilder.UserType);
+            var emailTokenProviderType = typeof(EmailTokenProvider<>).MakeGenericType(identityBuilder.UserType);
+            var authenticatorProviderType = typeof(AuthenticatorTokenProvider<>).MakeGenericType(identityBuilder.UserType);
+
+            identityBuilder.AddTokenProvider(TokenOptions.DefaultProvider, dataProtectionProviderType)
+                .AddTokenProvider(TokenOptions.DefaultEmailProvider, emailTokenProviderType)
+                // .AddTokenProvider(TokenOptions.DefaultPhoneProvider, phoneNumberProviderType)
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/Startup.cs:229]: Reminder to file an issue to uncomment and add it in the provider

**Review by sebastienros — APPROVED** [review]

File: src/OrchardCore.Modules/OrchardCore.Users/Startup.cs (hunk 3 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/Startup.cs:229]: created https://github.com/OrchardCMS/OrchardCore/issues/13710

**Comment by sebastienros** [comment]
Looking at aspnetcore I don't think the user id/name is ever logged.
Example: https://github.com/dotnet/aspnetcore/blob/8aa4a527c5d0393c6d8cfc4b7f05224aae93d3f4/src/Identity/UI/src/Areas/Identity/Pages/V5/Account/Logout.cshtml.cs

**Comment by MikeAlhayek** [comment]
I think I was looking at a code that was generated using the code generator. Either way, it be good news is that we don't log that private info also.

**Pull request merged**